File: analysis/risu_ghidra_scripts/scripts/ExportInterestingFunctions.py
# PyGhidra post-processing script. Run against the saved project with analyze=False.
from pathlib import Path
import re

from ghidra.app.decompiler import DecompInterface
from ghidra.program.model.data import StringDataInstance

NEEDLES = (
    "login", "auth", "token", "party", "invite", "map", "portal", "move",
    "battle", "auto", "npc", "monster", "boss", "quest", "pet", "proxy",
    "allocate", "free?port", "socks", "risu", "flash", "character", "line",
)

out = Path(str(getScriptArgs()[0]))
funcs_dir = out / "functions"
funcs_dir.mkdir(parents=True, exist_ok=True)

listing = currentProgram.getListing()
function_manager = currentProgram.getFunctionManager()
reference_manager = currentProgram.getReferenceManager()
selected = {}

def clean(value):
    return str(value).replace("\t", " ").replace("\r", " ").replace("\n", " ")

with (out / "strings-xrefs.tsv").open("w", encoding="utf-8", newline="\n") as stream:
    stream.write("string_address\tfunction_address\tfunction_name\tstring\n")
    iterator = listing.getDefinedData(True)
    while iterator.hasNext() and not monitor.isCancelled():
        item = iterator.next()
        if not item.hasStringValue():
            continue
        try:
            value = StringDataInstance.getStringDataInstance(item).getStringValue()
        except Exception:
            continue
        if value is None or not any(needle in str(value).lower() for needle in NEEDLES):
            continue
        # NativeAOT code commonly references the managed string object/header rather
        # than the first character address identified by Ghidra. Probe nearby headers.
        references = []
        seen_refs = set()
        for delta in (0, -4, -8, -12, -16):
            probe = item.getAddress().add(delta)
            for reference in reference_manager.getReferencesTo(probe):
                key = (str(reference.getFromAddress()), str(reference.getToAddress()))
                if key not in seen_refs:
                    seen_refs.add(key)
                    references.append(reference)
        if not references:
            stream.write(f"{item.getAddress()}\t\t\t{clean(value)}\n")
        for reference in references:
            function = function_manager.getFunctionContaining(reference.getFromAddress())
            if function is not None:
                selected[str(function.getEntryPoint())] = function
            faddr = "" if function is None else str(function.getEntryPoint())
            fname = "" if function is None else clean(function.getName())
            stream.write(f"{item.getAddress()}\t{faddr}\t{fname}\t{clean(value)}\n")

decompiler = DecompInterface()
decompiler.toggleCCode(True)
decompiler.toggleSyntaxTree(True)
decompiler.setSimplificationStyle("decompile")
if not decompiler.openProgram(currentProgram):
    raise RuntimeError("Could not open program in decompiler")

try:
    with (out / "functions-index.tsv").open("w", encoding="utf-8", newline="\n") as index:
        index.write("address\tname\tfile\tstatus\n")
        for address in sorted(selected):
            if monitor.isCancelled():
                break
            function = selected[address]
            safe_name = re.sub(r"[^A-Za-z0-9_.-]", "_", str(function.getName()))
            filename = f"{address}_{safe_name}.c"
            result = decompiler.decompileFunction(function, 90, monitor)
            status = "ok" if result.decompileCompleted() else clean(result.getErrorMessage())
            index.write(f"{address}\t{clean(function.getName())}\t{filename}\t{status}\n")
            with (funcs_dir / filename).open("w", encoding="utf-8", newline="\n") as source:
                source.write(f"/* Address: {address} | Ghidra name: {function.getName()} */\n")
                if result.decompileCompleted() and result.getDecompiledFunction() is not None:
                    source.write(str(result.getDecompiledFunction().getC()))
                else:
                    source.write(f"/* Decompile failed: {status} */\n")
finally:
    decompiler.dispose()

print(f"Exported {len(selected)} interesting functions to {out}")

File: analysis/risu_ghidra_scripts/scripts/ExportInterestingFunctions.java
// Ghidra headless post-script: export strings and decompile functions that reference
// application-oriented strings. This is an evidence collector, not a source restorer.
import ghidra.app.decompiler.*;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.data.StringDataInstance;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.Reference;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class ExportInterestingFunctions extends GhidraScript {
    private static final String[] NEEDLES = {
        "login", "auth", "token", "party", "invite", "map", "portal", "move",
        "battle", "auto", "npc", "monster", "boss", "quest", "pet", "proxy",
        "allocate", "free?port", "socks", "risu", "flash", "character", "line"
    };

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length != 1) throw new IllegalArgumentException("Expected output directory");
        File out = new File(args[0]);
        File funcsDir = new File(out, "functions");
        funcsDir.mkdirs();

        Set<Function> selected = new TreeSet<>(Comparator.comparing(f -> f.getEntryPoint().toString()));
        try (PrintWriter strings = writer(new File(out, "strings-xrefs.tsv"))) {
            strings.println("string_address\tfunction_address\tfunction_name\tstring");
            DataIterator data = currentProgram.getListing().getDefinedData(true);
            while (data.hasNext() && !monitor.isCancelled()) {
                Data item = data.next();
                if (!item.hasStringValue()) continue;
                String value;
                try { value = StringDataInstance.getStringDataInstance(item).getStringValue(); }
                catch (Exception ignored) { continue; }
                if (value == null || !interesting(value)) continue;
                Reference[] refs = getReferencesTo(item.getAddress());
                if (refs.length == 0) strings.println(tsv(item.getAddress(), null, null, value));
                for (Reference ref : refs) {
                    Function f = getFunctionContaining(ref.getFromAddress());
                    if (f != null) selected.add(f);
                    strings.println(tsv(item.getAddress(), f == null ? null : f.getEntryPoint(),
                        f == null ? null : f.getName(), value));
                }
            }
        }

        DecompInterface decompiler = new DecompInterface();
        decompiler.toggleCCode(true);
        decompiler.toggleSyntaxTree(true);
        decompiler.setSimplificationStyle("decompile");
        if (!decompiler.openProgram(currentProgram)) throw new IOException("Cannot open program in decompiler");

        try (PrintWriter index = writer(new File(out, "functions-index.tsv"))) {
            index.println("address\tname\tfile\tstatus");
            for (Function f : selected) {
                if (monitor.isCancelled()) break;
                String base = f.getEntryPoint() + "_" + safe(f.getName()) + ".c";
                DecompileResults result = decompiler.decompileFunction(f, 90, monitor);
                String status = result.decompileCompleted() ? "ok" : result.getErrorMessage();
                index.println(f.getEntryPoint() + "\t" + clean(f.getName()) + "\t" + base + "\t" + clean(status));
                try (PrintWriter pw = writer(new File(funcsDir, base))) {
                    pw.println("/* Address: " + f.getEntryPoint() + " | Ghidra name: " + f.getName() + " */");
                    if (result.decompileCompleted() && result.getDecompiledFunction() != null)
                        pw.println(result.getDecompiledFunction().getC());
                    else pw.println("/* Decompile failed: " + clean(status) + " */");
                }
            }
        } finally {
            decompiler.dispose();
        }
        println("Exported " + selected.size() + " interesting functions to " + out.getAbsolutePath());
    }

    private boolean interesting(String value) {
        String lower = value.toLowerCase(Locale.ROOT);
        for (String needle : NEEDLES) if (lower.contains(needle)) return true;
        return false;
    }

    private PrintWriter writer(File file) throws IOException {
        return new PrintWriter(new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8));
    }

    private String tsv(Address stringAddr, Address functionAddr, String functionName, String value) {
        return stringAddr + "\t" + (functionAddr == null ? "" : functionAddr) + "\t" +
            clean(functionName) + "\t" + clean(value);
    }

    private String clean(String s) { return s == null ? "" : s.replace('\t', ' ').replace('\r', ' ').replace('\n', ' '); }
    private String safe(String s) { return clean(s).replaceAll("[^A-Za-z0-9_.-]", "_"); }
}
